src/index.ts

```diff
@@ -9,65 +10,145 @@ class VPCPlugin {
+    if (!vpc || vpc.vpcName == null || (vpc.subnetNames == null && vpc.securityGroupNames == null)) {
+      throw new Error(
+        "Serverless file is not configured correctly. " +
+        "You must specify the vpcName and at least one of subnetNames or securityGroupNames. " +
+        "Please see README for proper setup."
+      );
     }
   }
 
   /**
    * Setup AWS resources
    */
   public initAWSResources (): void {
-    const credentials = this.serverless.providers.aws.getCredentials()
-    credentials.region = this.serverless.providers.aws.getRegion()
-    this.ec2Wrapper = new EC2Wrapper(credentials)
+    const credentials = this.serverless.providers.aws.getCredentials();
+    credentials.region = this.serverless.providers.aws.getRegion();
+    this.ec2Wrapper = new EC2Wrapper(credentials);
   }
 
   /**
-   * Gets the desired vpc with the designated subnets and security groups
-   * that were set in serverless config file
+   * Updates functions vpc config
    * @returns {Promise<object>}
    */
-  public async updateVpcConfig (): Promise<object> {
-    this.serverless.cli.log("Updating VPC config...")
+  public async updateFunctionsVpcConfig (): Promise<object> {
+    Globals.logInfo("Updating VPC config...");
+    const service = this.serverless.service;
+
+    // Sets the serverless's vpc config
+    if (service.functions) {
+      // get basic VPC config
+      const basicVPCDiscovery = await this.getVpcConfig(service.custom.vpcDiscovery);
+      // loop through the functions and update VPC config
+      for (const fName of Object.keys(service.functions)) {
+        const f = service.functions[fName];
+        const fVPCDiscovery = f.vpcDiscovery;
+
+        if (typeof fVPCDiscovery === "boolean" && !fVPCDiscovery) {
+          // skip vpc setup for `vpcDiscovery=false` option
+          Globals.logInfo(`Skipping VPC config for the function '${fName}'`);
+          continue;
+        }
+
+        const vpcDiscovery = basicVPCDiscovery;
+        // check vpcDiscovery config
+        if (fVPCDiscovery) {
+          const noSubnetsAndGroups = fVPCDiscovery.subnetNames == null && fVPCDiscovery.securityGroupNames == null;
+          // validate vpcDiscovery config options
+          if (fVPCDiscovery.vpcName == null || noSubnetsAndGroups) {
+            Globals.logWarning(
+              `The function '${fName}' is not configured correctly.` +
+              "Please see README for proper setup. The basic vpc config are applied"
+            );
+          } else {
+            // merge basic config to specific for the function
+            Object.assign(vpcDiscovery, await this.getVpcConfig(fVPCDiscovery));
+          }
+        }
+
+        // init vpc empty config in case not exists
+        f.vpc = f.vpc || {};
+
+        // log warning in case vpc.subnetIds and vpcDiscovery.subnetNames are specified.
+        if (f.vpc.subnetIds && fVPCDiscovery.subnetNames) {
```

Comment:
We probably should have a README section to describe the behavior in all of these permutations. It might not be obvious to a user how `vpc` and `vpcDiscovery` options interact